### ISSUE TITLE
svg fill-opacity [Bug]

### ISSUE THREAD

**Issue by detrothe** [jonobr1/two.js#710]
Hi,
I create a technical drawing on screen that also needs to be printed. I use svg. The result is as expected on screen but if I take the svg content and save it on disk (as demonstrated in the examples), and open in in inkscape, libreoffice or microsoft office 2019 (no difference between ubuntu and windows), then the area inside the path is black. This happens with my own moment arrow (left) as well as with your makeArrow (right side).
my code in Javascript for the arc segmentt: 

       var vertices = [];
        dalpha = Math.PI / (teilung + 1)
        alpha = 0.0
        for (let i = 0; i < teilung - 2; i++) {
            x = tr.Pix0(-radius * Math.sin(alpha))
            z = tr.Pix0(radius * Math.cos(alpha))
            vertices.push(new Two.Anchor(x, z));
            alpha += dalpha
        }
        let curve = new Two.Path(vertices, false, true)
        curve.linewidth = linewidth;
        curve.stroke = color;
        curve.noFill()

        group.add(curve)
        // arrowhead  follows

I can fix the problem by hand,if I exchange in the svg-file in both cases:  fill-opacity="1"  to fill-opacity="0". 
fill="transparent" in the svg-file seems to have no impact and I have not found a hint that this is part of the specs. But I'm not an expert on svg.

Result on screen:
![screen](https://github.com/jonobr1/two.js/assets/98234093/938f2b8f-f77b-4079-a816-e11d42292da3)

in inkscape without any changes:
![generated-by-two-js](https://github.com/jonobr1/two.js/assets/98234093/0d1e05b9-35aa-4661-905a-83cb2a9f58c1)

corrected with fill-opacity="0"
![modified](https://github.com/jonobr1/two.js/assets/98234093/0cb2e8b7-7a65-4c64-8b85-778a14bb7341)

Additional question: Is there a method in Two.Path to use fill-opacity?

Thank you, I like your library

**Comment by jonobr1** [comment]
Have you tried `curve.fill = "none";`?

**Comment by detrothe** [comment]
Hi,

if I use curve.fill="none", there is still transparent in the svg-file.
On screen it works like nofill().

But if I edit the svg-file to fill="none", then in all cases ( inkscape,
libre office and microsoft office 2019 ) the filling is gone.  Hope you
can fix it easily.

I would also need a npm version for netlify. Sorry.

Detlef


Am 09.10.23 um 21:55 schrieb Jono:

**Comment by jonobr1** [comment]
Unfortunately different browsers have different implementations of `"none"`, so it's not a great option for Two.js.

Have you tried `curve.fill = "rgba(0, 0, 0, 0);"`?

**Comment by detrothe** [comment]
Hi,

rgba doesn't work in inkscape, ... Only curve.fill="none" does the job.

I have tested it on following browsers: chrome (chromium), Firefox, iOS
Safari 16.6 on old Apple tablett, Firefox and google chrome on Android

I looked at the following page: https://www.w3.org/TR/SVG2/painting.html

They only use fill="none" in all examples, no transparent or rgba is
mentioned.


Am 10.10.23 um 00:40 schrieb Jono:

**Comment by jonobr1** [comment]
Got it. That makes sense. If this is a static image (not animating) and you are using the SVG Renderer then you can always access the underlying SVG element in code like this. But, be sure to call `two.update();` or have the renderer render at least once beforehand so that the elements are created:

```javascript
// Given your curve code
two.add(curve);
two.update();

const elem = curve.renderer.domElement;
elem.setAttribute('fill-opacity', 0);
```

**Comment by detrothe** [comment]
It is a static image. The image is created after a finite element calculation, and two.update() is always called. Your suggestion with curve.fill="none" serves me fine. I don't have to care about old browsers, because my students have current hardware, usually newer than mine.
But two.makeArrow still has black background, see my first comment, (picture in the middle). Other users of your lib might have that problem too and .nofill() will not work as expected . Or am I wrong?

**Comment by jonobr1** [comment]
Got it. I'll have to re-run tests to see if all the browsers support `curve.fill = "none"`. At the time of writing, which was a long time ago, both Firefox and Safari did not support `"none"`. They may now.

**Comment by jonobr1** [comment]
No, because `fill-opacity` is not available in all renderers. That is why it's recommended to use:

```javascript
curve.renderer.domElement.setAttribute('fill-opacity', 0);
```

**Comment by detrothe** [comment]
ok. I have one program using three.js and there I use for example;

const matLite = new THREE.MeshBasicMaterial({
            color: color,
            transparent: true,
            opacity: 0.4,
            side: THREE.DoubleSide
        });

I haven't used canvas, but there must be something. Even 30 years old shapes in excel VBA have that. 
But its probably a bigger thing to implement a general function.

**Comment by jonobr1** [comment]
v8.12 will have noFill and noStroke execute CSS commands `"none"` instead of `"transparent"`

**Comment by detrothe** [comment]
Thank you very much for the changes.--Diese Nachricht wurde von meinem Android Mobiltelefon mit WEB.DE Mail gesendet.Am 16.10.23, 20:00 schrieb Jono ***@***.***>:

   
  Closed #710 as completed.
